src/analysermodel.cpp

```diff
@@ -348,4 +349,21 @@ bool AnalyserModel::needAcothFunction() const
     return mPimpl->mNeedAcothFunction;
 }
 
+bool AnalyserModel::areEquivalentVariables(const VariablePtr &variable1,
+                                           const VariablePtr &variable2)
+{
+    auto key = reinterpret_cast<intptr_t>(variable1.get()) * reinterpret_cast<intptr_t>(variable2.get());
+    auto cacheKey = mPimpl->mCachedEquivalentVariables.find(key);
+
```

Comment:
Do I have this correct: You're keeping track of pairs of variables by storing the product of their pointers, after the pointers have been cast to integers? How can that be guaranteed to be unique? Most of the time, sure ... but it's a bit weird? How is it faster than caching some other format that would be unique (and a lot cleaner) like a binned combination of indices?  And if you do want to keep it like this, why does it need to be a signed integer pointer cf. unsigned?  I'm so confused ...

Reply:
You are correct, we currently use as a key the product of the address of both variables. There is indeed no guarantee of uniqueness. Also, the casting to `intptr_t` is because it's the type that should be used for addresses. That type will increase the risk of non-uniqueness, but should indeed be easily overcome by using another type (even a `double`).

Now, regarding the product of addresses, the idea is to come up with a very fast way of computing the key. I originally had a string of the formwat `<@variable1><@variable2>`, but it was too time consuming to generate (more than 2.5x slower to run the `Generator.fabbriFantiniWildersSeveriHumanSanModel2017` test). FWIW, this is how I was creating that string-based key:
```c++
std::ostringstream strs;
strs << variable1.get() << variable2.get();
auto key = strs.str();
```
rather than:
```c++
auto key = reinterpret_cast<intptr_t>(variable1.get()) * reinterpret_cast<intptr_t>(variable2.get());
```

So, what do you mean exactly by "binned combination of indices"? I am certainly not happy with the product of addresses, so I would be happy to go with another solution as long as it is (roughly) as fast.

Reply:
Have you thought about a sparse symmetric matrix, where you have an entry only in places where there is an equivalence between variables?  (Depending on how you decided to code the rows/columns it wouldn't necessarily need to be sparse either).  The coordinates of the matrix rows/columns could be an indexed list of the pointers already processed.  Since you already know that models passed to the `Analyser` are valid, you can use a simpler way than the pointer addresses to represent them: component and variable indices, for example? Using the pointers' addresses as if they are strings makes for something that's probably not going to be amenable to fast searching through the map structure.
I'm not sure about how speed would scale - that would need testing - but at least it could guarantee uniqueness.  
It could perhaps be used as a cleaner/faster way to store equivalences outside the analyser too (though I guess the Analyser is the object that "needs" it the most, so ... dunno ).

Reply:
I went for a much simpler approach (see commit fd0be4408) which relies on the [Cantor pairing function](https://en.wikipedia.org/wiki/Pairing_function#Cantor_pairing_function).

Either way, thanks for "insisting" on the non-uniqueness of my previous key implementation. I should have been more diligent.